crypto/rand/rand_lib.c

```diff
@@ -193,6 +193,73 @@ size_t rand_drbg_get_entropy(RAND_DRBG *drbg,
     return ret;
 }
 
+/*
+ * Find a suitable system time.  Start with the highest resolution source
+ * and work down to the slower ones.  This is added as additional data and
+ * isn't counted as randomness, so any result is acceptable.
+ */
+static uint32_t get_timer_bits(void)
+{
+    uint32_t res = OPENSSL_rdtsc();
+
+    if (res != 0)
+        return res;
+#if defined(_WIN32)
+    {
+        LARGE_INTEGER t;
+        FILETIME ft;
+
+        if (QueryPerformanceCounter(&t) != 0)
+            return (uint32_t)t.QuadPart;
+        GetSystemTimeAsFileTime(&ft);
+        return (uint32_t)ft.dwLowDateTime;
+    }
+#elif defined(__sun) || defined(__hpux)
+    return (uint32_t)gethrtime();
+#else
+# if defined(_AIX)
+    {
+        timebasestruct_t t;
+
+        read_wall_time(&t, TIMEBASE_SZ);
+        if (time_base_to_time(&t, TIMEBASE_SZ) == 0)
+            return (uint32_t)t.tb_low;
+    }
+# endif
+
+# if defined(_POSIX_C_SOURCE) \
+     && defined(_POSIX_TIMERS) \
+     && _POSIX_C_SOURCE >= 199309L \
+     && (!defined(__GLIBC__) || __GLIBC_PREREQ(2, 17))
+    {
+        struct timespec ts;
+        clockid_t cid;
+
+#  ifdef CLOCK_BOOTTIME
+        cid = CLOCK_BOOTTIME;
+#  elif defined(_POSIX_MONOTONIC_CLOCK)
+        cid = CLOCK_MONOTONIC;
+#  else
+        cid = CLOCK_REALTIME;
+#  endif
+
+        if (clock_gettime(cid, &ts) == 0)
+            return (uint32_t)ts.tv_nsec;
+    }
+# endif
+# if defined(__unix__) \
+     || (defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 200112L)
+    {
+        struct timeval tv;
+
+        if (gettimeofday(&tv, NULL) == 0)
+            return (uint32_t)tv.tv_usec;
+    }
+# endif
+    return (uint32_t)time(NULL);
+#endif
+}
```

Comment:
All the explicit casting to `uint32_t` in this function worries me.  Why is it needed?

Reply:
There was a discussion a while ago in #4757 because `OPENSSL_rdtsc` did not have a central prototype and was declared inconsistently. For some reason which I don't recall, it was settled down that the return type should be `uint32_t`.  This made me wonder that time, since according to [Wikipedia](https://en.wikipedia.org/wiki/Time_Stamp_Counter) its a 64bit value even on x86 architectures. (Sorry for not providing an Intel manual page.) But I didn't pursue that any further.

Reply:
It looks like we are spinning in circles and that @dot-asm tries to tell us since November that we should dump this profiling code, which is an obsolete debugging artifact.

See https://github.com/openssl/openssl/pull/4757#issuecomment-346038372

and here https://github.com/openssl/openssl/pull/5231#discussion_r165612740

and https://github.com/openssl/openssl/pull/5231/files#r165923037

If we remove the code, then probably the only use case for the rdtsc instruction that remains is the additional input for the drbg...

Reply:
See

```C
msp@msppc:~/src/openssl$ grep -R 'rdtsc('
crypto/armcap.c:27:uint32_t OPENSSL_rdtsc(void)
crypto/armcap.c:51:uint32_t OPENSSL_rdtsc(void)
crypto/rand/rand_lib.c:61:            c = (unsigned char)(OPENSSL_rdtsc() & 0xFF);
crypto/rand/rand_lib.c:240:    tsc = OPENSSL_rdtsc();
crypto/sparcv9cap.c:102:uint32_t OPENSSL_rdtsc(void)
include/internal/cryptlib.h:98:uint32_t OPENSSL_rdtsc();
test/siphash_internal_test.c:58:    stopwatch = OPENSSL_rdtsc();
test/siphash_internal_test.c:61:    stopwatch = OPENSSL_rdtsc() - stopwatch;
test/siphash_internal_test.c:65:    stopwatch = OPENSSL_rdtsc();
test/siphash_internal_test.c:71:    stopwatch = OPENSSL_rdtsc() - stopwatch;
test/poly1305_internal_test.c:55:    stopwatch = OPENSSL_rdtsc();
test/poly1305_internal_test.c:58:    stopwatch = OPENSSL_rdtsc() - stopwatch;
test/poly1305_internal_test.c:62:    stopwatch = OPENSSL_rdtsc();
test/poly1305_internal_test.c:68:    stopwatch = OPENSSL_rdtsc() - stopwatch;
test/modes_internal_test.c:904:    start = OPENSSL_rdtsc();
test/modes_internal_test.c:906:    gcm_t = OPENSSL_rdtsc() - start;
test/modes_internal_test.c:911:    start = OPENSSL_rdtsc();
test/modes_internal_test.c:915:    ctr_t = OPENSSL_rdtsc() - start;
test/modes_internal_test.c:927:        start = OPENSSL_rdtsc();
test/modes_internal_test.c:930:        gcm_t = OPENSSL_rdtsc() - start;
```

Reply:
This is `get_timer_bits`, there's absolutely no reason `res` *must* be a `uint32_t` just because that's what `OPENSSL_rdtsc` returns.  `res` could as well be a `uint64_t` and accomodate everything without explicit casts, as far as I know, no?

Reply:
The debugging artifacts isn't about this function, but about the affected `test/*internal*.c`

Reply:
Agreed, `get_timer_bits` could and should return `uint64_t`.

As for `OPENSSL_rdtsc`:  I just wanted to say that it is not clear to me why (1) it returns a `uint32_t` in the first place when the RDTSC instruction actually returns 64bit, and why (2) we spend so much time fixing the `*_internal_test.c` profiling code when Andy says that it should be removed anyway.

Reply:
> The debugging artifacts isn't about this function, but about the affected test/*internal*.c

Sorry, it was misleading that I posted my reply directly to your question in the review comment. I should have posted it as general comment.

Reply:
See #5268 for removal

Reply:
I'll undo the profiling code changes -- I kept them as a separate commit just in case.

The reason `OPENSSL_rdstc` returns an `uint32_t` is because on some architectures the equivalent register is only thirty two bits in size.  Declaring it to return thirty two bits worked for all platforms, declaring it to return sixty four meant rubbish on some.  Still, I'm happy to change the result of `get_timer_bits` to be ab `uint64_t`.